pipeline/detection.py

```diff
@@ -98,20 +201,422 @@ def run(self, *args, **kwargs):
+        tmpsources = tmpimage.parent / f'{tmpnamebase}.sources.fits'
+        tmpparams = tmpimage.parent / f'{tmpnamebase}.param'
+
+        # For debugging purposes
+        self._tmpimage = tmpimage
+        self._tmpweight = tmpweight
+        self._tmpflags = tmpflags
+        self._tmpsources = tmpsources
+
+        if image.data is None or image.weight is None or image.flags is None:
+            raise RuntimeError( f"Must have all of image data, weight, and flags" )
+
+        # Figure out where astromatic config files are:
+        astromatic_dir = None
+        cfg = Config.get()
+        if cfg.value( 'astromatic.config_dir' ) is not None:
+            astromatic_dir = pathlib.Path( cfg.value( 'astromatic.config_dir' ) )
+        elif cfg.value( 'astromatic.config_subdir' ) is not None:
+            astromatic_dir = pathlib.Path( CODE_ROOT ) / cfg.value( 'astromatic.config_subdir' )
+        if astromatic_dir is None:
+            raise FileNotFoundError( "Can't figure out where astromatic config directory is" )
+        if not astromatic_dir.is_dir():
+            raise FileNotFoundError( f"Astromatic config dir {str(astromatic_dir)} doesn't exist "
+                                     f"or isn't a directory." )
+
+        # TODO : make these configurable by instrument (at least!)
+        # For now we're using the astromatic defaults that everybody
+        # just uses....
+        conv = astromatic_dir / "default.conv"
+        nnw = astromatic_dir / "default.nnw"
+
+        if not ( conv.is_file() and nnw.is_file() ):
+            raise FileNotFoundError( f"Can't find SExtractor conv and/or nnw file: {conv} , {nnw}" )
+
+        if psffile is not None:
+            if not pathlib.Path(psffile).is_file():
+                raise FileNotFoundError( f"Can't read PSF file {psffile}" )
+            psfargs = [ '-PSF_NAME', psffile ]
+            paramfilebase = astromatic_dir / "sourcelist_sextractor_with_psf.param"
+        else:
+            psfargs = []
+            paramfilebase = astromatic_dir / "sourcelist_sextractor.param"
+
+        # SExtractor reads the measurements it produes from a parmeters
+        # file.  We need to edit it, though, so that the number of
+        # apertures we have matches the apertures we ask for.
+        # TODO : review the default param file and make sure we have the
+        # things we want, and don't have too much.
+        if len(apers) == 1:
+            paramfile = paramfilebase
+        else:
+            paramfile = tmpparams
+            with open( paramfilebase ) as ifp:
+                params = [ line.strip() for line in ifp.readlines() ]
+            for i in range(len(params)):
+                if params[i] in [ "FLUX_APER", "FLUXERR_APER" ]:
+                    params[i] = f"{params[i]}({len(apers)})"
+            with open( paramfile, "w") as ofp:
+                for param in params:
+                    ofp.write( f"{param}\n" )
```

Comment:
So much clunkiness I can't take it anymore :exploding_head:

Reply:
Think about it like this.

Whenever you use a library, you have to interface to that library.  If it's doing anything that's non-trivial, that interface may make decisions and assumptions that differ from what you would have.  This kind of thing happens with Pandas, for instance, where sometimes to import things you have to wrangle the data a bit.  In astronomy, I've seen various libraries take weight (inverse variance) images, inverse noise images, noise images, and variance images.  (SExtractor lets you tell it which one you're passing, but others just demand a certain thing.) In that case, you'd have to reformat your data, even if you're just making a library call and not running a subprocess.  All of this is a little bit clunky, but it's the price you pay for using libraries that don't do everything exactly the way you might have wanted to, or that don't do everything in a way that perfectly matches your current use case.

These external programs we're calling just use reading and writing temporary files, and subprocess.run(), as part of their interface.  It's a little clunkier than just a function call passing a whole bunch of parameters, but not *that* much, given that you might have had to adapt your data for what the function's parameters wanted anyway.

Right here, we could make this a little bit less clunky by not having a default parameters file at all, and just keeping a list of the parameters we want as part of the code, and writing and deleting the temporary parameters file every time.  I considered doing that, but decided not to, because for debugging purposes it was convenient to have a the base parameters file just sitting there ready to use.

Reply:
I guess you are right. At least we don't have to build a robot arm to take floppy disks in and out of an old drive "because that's how the software runs" back in the early 90s...

Reply:
...of course, when we get to writing to tape on HPSS, there *will be* robot arms involed...